src/gpio.rs

```diff
@@ -127,6 +136,127 @@ pub struct Disabled;
 
 pub struct Unknown;
 
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+pub struct SubscribedInput;
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+struct UnsafeCallback(*mut Box<dyn FnMut() + 'static>);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+impl UnsafeCallback {
+    #[allow(clippy::type_complexity)]
+    pub fn from(boxed: &mut Box<Box<dyn FnMut() + 'static>>) -> Self {
+        Self(boxed.as_mut())
+    }
+
+    pub unsafe fn from_ptr(ptr: *mut c_types::c_void) -> Self {
+        Self(ptr as *mut _)
+    }
+
+    pub fn as_ptr(&self) -> *mut c_types::c_void {
+        self.0 as *mut _
+    }
+
+    pub unsafe fn call(&mut self) {
+        let reference = self.0.as_mut().unwrap();
+
+        (reference)();
+    }
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+static ISR_SERVICE_ENABLED: core::sync::atomic::AtomicBool =
+    core::sync::atomic::AtomicBool::new(false);
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+unsafe extern "C" fn irq_handler(unsafe_callback: *mut esp_idf_sys::c_types::c_void) {
+    let mut unsafe_callback = UnsafeCallback::from_ptr(unsafe_callback);
+    unsafe_callback.call();
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+fn enable_isr_service() -> Result<(), EspError> {
+    if ISR_SERVICE_ENABLED.compare_exchange(
+        false,
+        true,
+        core::sync::atomic::Ordering::SeqCst,
+        core::sync::atomic::Ordering::Relaxed,
+    ) == Ok(false)
+    {
+        if let Err(e) = esp!(unsafe { esp_idf_sys::gpio_install_isr_service(0) }) {
+            ISR_SERVICE_ENABLED.store(false, core::sync::atomic::Ordering::SeqCst);
+            return Err(e);
+        }
+    }
+    Ok(())
+}
+
+#[cfg(all(not(feature = "riscv-ulp-hal"), feature = "alloc"))]
+type ClosureBox = Box<Box<dyn FnMut()>>;
```

Comment:
With this alias, users are forced to allocate. However if `PinNotifySubscription` is generic on the inner box then allocation could (when the closure is a free function) be avoided completely. i.e. `pub(crate) struct PinNotifySubscription<T: FnMut()>(i32, Box<T>);`

See https://gist.github.com/Dominaezzz/cb1695ac34618fb46dcffd67f314c281

Reply:
In your example, it seems that nothing prevents the user from creating more than one service with different `intr_alloc_flags` which would cause undefined behavior. Maybe the best solution would be to ditch the Atomic and replace this with a proper Mutex.

I don't understand what you mean "the users are forced to allocate". It seems to me that in any case you need to allocate by Boxing the closure. What is exactly gained with this generic ?

Reply:
If the closure is a free function, the resulting type is a ZST. Box does not allocate if the type is zero sized.
So `Box<fn()>` is free.

Reply:
Ok, thanks for the answer and your remarks. I will give a try to this implementation.

Reply:
Thanks for your comments @Dominaezzz. 

In 28830cf9d08d1f647a460386721374edb2393b45 I tried to implement something in agreement with your proposal around the `Service` which is used to make sure the isr_service is installed with `gpio_install_isr_service`.
However I kept the double boxing and introduced an Rc to wrap the `Service`.

If I understood correctly the struct used for TypeState should be zero sized. But we need to keep some struct around to make sure:
- The closure is still around while the handler is registered
- The `Service` is not being dropped while a Pin is in `Subscribed` state

I chose to keep this in a static array, which I don't know how could be used with the generic you proposed.

@ivmarkov @MabezDev I'd like to have your opinion on this as it changes the way users would use the subscription. You can check an example usage [here](https://github.com/pyaillet/esp32-isr/blob/use-patched-eih/src/main.rs#L71).

Reply:
I can see why the nested box couldn't be avoided and the Rc is needed. It's a consequence of not being able to return any state to the user, since all that's returned is the ZST pin state machine.

My only concern now is how I'm supposed to use this as driver writer. Normally I'd have a method like, `Driver::new(sclk: impl InputPin, busy: impl InputPin, rst: impl OutputPin): Result<Self, .....>`.
If my driver needs to setup an interrupt, how would I do that with this API? Would I accept a service and a `Gpio<...>` and then transform it into a subscribed pin? Or should I have the caller do this, since they've been doing stuff like `into_output()`/`into_input()` before calling the driver ctor.

Reply:
I think that defining this abstraction won't be an easy task and was not really the goal here.

It will depend on the driver needs, the platform capabilities, ...

Reply:
> I can see why the nested box couldn't be avoided and the Rc is needed. It's a consequence of not being able to return any state to the user, since all that's returned is the ZST pin state machine.
> 
> My only concern now is how I'm supposed to use this as driver writer. Normally I'd have a method like, `Driver::new(sclk: impl InputPin, busy: impl InputPin, rst: impl OutputPin): Result<Self, .....>`. If my driver needs to setup an interrupt, how would I do that with this API? Would I accept a service and a `Gpio<...>` and then transform it into a subscribed pin? Or should I have the caller do this, since they've been doing stuff like `into_output()`/`into_input()` before calling the driver ctor.

@Dominaezzz Obviously it is impossible with this API to implement a higher level "driver" which is MCU-independent (btw, I guess by "driver" you mean a piece of code modelling something like a button etc. sitting on top of the GPIO API). `embedded-hal` is currently simply missing traits that model the notion of "subscription" to a pin, or the notion of an "interrupt". Now, we can try implementing such traits / abstractions ourselves, but I don't see this going anywhere. Such an exercise would be useful only if we expect long term to see similar traits popping up in `embedded-hal` and then our traits would be retired/migrated/whatever to theirs, but I'm very sceptical here (see why further below). 

So to answer your question - your "driver" will have to rely on the ESP32-specific subscription API currently, and use the ESP32 generified pin type (OR - the down-graded `GpioPin<STATE>` struct instead). And if you want to have it MCU agnostic, you'll have to build your own abstractions on top of that.

Future of `embedded-hal`, interrupts, subscriptions etc.:
============================================

I think where `embedded-hal` is going is into async for a lot of stuff, [including gpio-based interrupts, which are expressed very cleanly in the async world](https://github.com/rust-embedded/embedded-hal/blob/master/embedded-hal-async/src/digital.rs#L33).

Basically, you don't need subscriptions, callback functions etc. If your pin implements the async e-hal `Wait` trait (perhaps, in addition to the sync e-hal `InputPin` trait which is anyway async-compatible in that it does not block), you would just - in your async code - "wait" on low, high or edge, but this wait would be non-blocking, so your app would continue to do other stuff in the meantime.

So why are we exposing a subscription based ESP32-specific interface then?
a) Because this is what ESP IDF has anyway, but we want to make it a bit more ergonomic, with less unsafe
b) Because later on, once the `embedded-hal-async` hits 0.1 or something, we can **implement** the async API on top of the subscription-based interface, as I was doing for other subscription based APIs in ESP IDF [here](https://github.com/esp-rs/esp-idf-svc/blob/master/src/eventloop.rs#L811) and [here](https://github.com/esp-rs/esp-idf-svc/blob/master/src/timer.rs#L153).
c) In the meantime, folks like me who are already using async can implement their own async wrappers on top of the subscription-based ESP32-specific API, even if they don't follow the `embedded-hal-async` traits which are not published yet.
d) Folks who would like to stay in the sync world can still use the subscription-based API, albeit in an ESP32-specific way.

Reply:
> @ivmarkov @MabezDev I'd like to have your opinion on this as it changes the way users would use the subscription. You can check an example usage [here](https://github.com/pyaillet/esp32-isr/blob/use-patched-eih/src/main.rs#L71).

Sorry for following up late here, but my week was busy. 
At a first sight, the implementation of an `IsrService` seems to elegantly bypass the whole problem of needing to synchronize the call to `gpio_install_isr_service`. However, it still relies on the fact that - in the presence of two simultaneous calls to `gpio_install_isr_service` - one would be succeeding, and the other would fail, *with proper cleanup after itself*. Now I'm reading, and re-reading the [C implementation]((https://github.com/espressif/esp-idf/blob/master/components/driver/gpio.c#L447)) of `gpio_install_isr_service` and I'm still not sure we are safe, without any explicit synchronization primitive (granted, atomic locks won't do), but [this one](https://github.com/esp-rs/esp-idf-hal/blob/master/src/mutex.rs#L13) might. 

Moreover, if we *are* safe without any synchronization primitive - we don't need the `IsrService` struct either. We can revert to @pyaillet's original implementation that did not use compare-exchange but just regular `load`/`save` atomic ops, let multiple threads try to initialize, and handle the case where `gpio_install_isr_service` is returning `ESP_ERR_INVALID_STATE` as a flag that the service IS already initialized by one of the competing threads.

As you've seen in sample code, having an `IsrService` hanging around is not so easy. For one, you are currently using `Rc` for shared access, but you should really be using `Arc`, as pins can be subscribed from several threads, so the `IsrService` should be then `Send` and `Sync` and `Clone`. Also, we end up having an `Rc` (which should be `Arc`) in our APIs, which is maybe not so nice (I mistake I did [elsewhere](https://github.com/esp-rs/esp-idf-svc/blob/master/src/wifi.rs#L247) that I still need to address / think about). So you might need to implement the whole thing upside-down, where the `Arc` is *inside* the IsrService, and the IsrService is itself `Send` + `Sync` + `Clone` + `'static`.

So the thing is - why bother? With `IsrService` we can also pass `intr_alloc_flags` and probably this is **the** thing that we should check if user needs control of:
- If yes, then I suggest we go with the `IsrService`, however the `Arc` should be probably *inside* it.
-  If no (and I hope it is a "no"), then I suggest we revert to the simpler model of *not* having an `IsrService` around

Reply:
I am really not sure of the implications of using other flags values for `gpio_install_isr_service`.
As far as I can tell however, [all examples provided](https://github.com/espressif/esp-idf/search?p=2&q=gpio_install_isr_service) in the codebase of the `espressif/esp-idf` project use `#define ESP_INTR_FLAG_DEFAULT 0`. So I would go with your second suggestion @ivmarkov

Reply:
Another thought: having an `IsrService` that the user can `new` just like that is a pattern we don't follow in `esp-idf-hal`. 

(We do follow that pattern in `esp-idf-svc` for stuff which looks like a driver, i.e. `EspWifi` and `EspEth`, but in retrospective, that was probably a mistake, and these in future might need to be moved to `esp-idf-hal` OR still live in `esp-idf-svc` but take a peripheral from `esp-idf-hal`; perhaps, we need a "modem" peripheral that lives in `esp-idf-hal` and is taken and owned by `EspWifi` in `esp-idf-svc` for the lifetime duration of `EspWifi`.)

Where I'm getting at is that if `IsrService` is to survive, perhaps it needs to take and own in its `new` method a `GpioIsr` peripheral, whose only instance is a member of the `Peripherals` structure [here](https://github.com/esp-rs/esp-idf-hal/blob/master/src/peripherals.rs#L77).

That would also solve the synchronization issues because then - by definition - calling `gpio_install_isr_service` can only be done by the `IsrService` instance that owns the `GpioIsr` peripheral. 

But then again, do we want this complexity (and the hidden `Arc`-ing it involves), or shall we hide `gpio_install_isr_service` completely and pretend that the notion of pin subscription is something each pin is responsible for, separately? Which means we **have to** have an answer whether there is something user would like to control in terms of flags passed to the `intr_alloc_flags` bitset of `gpio_install_isr_service`.

Reply:
> I am really not sure of the implications of using other flags values for `gpio_install_isr_service`. As far as I can tell however, [all examples provided](https://github.com/espressif/esp-idf/search?p=2&q=gpio_install_isr_service) in the codebase of the `espressif/esp-idf` project use `#define ESP_INTR_FLAG_DEFAULT 0`. So I would go with your second suggestion @ivmarkov

@pyaillet  OK so in that case how about the following:
- Let's revert - for now - to the original implementation that did not have the `IsrService` helper. If this turns out to be a mistake, well, we can restore `IsrService` later in the way I've described where it takes a `GpioIsr` "peripheral" in its `new` constructor that solves the concurrency issues, and it has a hidden `Arc` inside. 
- Protect the `gpio_install_isr_service` call either with a `esp_idf_hal::mutex::Mutex` instance, or with the atomic load/store checks that were there originally, plus a check whether the call to `gpio_install_isr_service` returned `ESP_ERR_INVALID_STATE` and assume that the call succeeded in another competing thread.

Reply:
I will do that.

For `gpio_install_isr_service` I would use a Mutex, if I am right it would make the second caller wait, but I guess this would occur in an initialization step which seems to be a minor penalty and more robust than relying on the return of `gpio_install_isr_service`.

Reply:
At the very least, if the service is going to be trapped in here, some sort of public `init` method should be exposed as I'll be using a different (platform agnostic) abstraction and I'd like some way to verify that a service has been created before I create a handler.
It'd be nice if the service was just exposed but I appreciate that `Rc`/`Arc` is less than ideal.

Although, the hidden Arc won't be necessary if the pin state is decoupled from handler registration. Since that would allow the generics I proposed to work.

Reply:
@Dominaezzz In my opinion you should not decouple handler registration (for a specific pin) from pin state because you could face an inconsistent state between the two.

I am not sure to understand your use case here, unless I'm missing something. The only things you lose are:
- The possibility to initialize the service with a different `ESP_INTR_FLAG`
- The possibility to remove *safely* the `gpio_isr_service`

The purpose of the current API is to make that you don't happen to have a handler registered without the service being installed.

Reply:
The inconsistent state could be addressed by having the handler keep a reference to the pin in the correct state.

In addition to losing those two things, there's now two `Box`es that I have to pay for and a `Mutex` I have to hit for every new subscription.

Without a HAL interface to comply to there's no right way to do this yet. I just ask for the `enable_isr_service()` to be public, so my library can still interop with this.

Reply:
I am not opposed to make `enable_isr_service()` public, but then again I don't understand what is the use.
The current API makes sure for you not to bother about this.

If we make this public:
- Another call to `enable_isr_service` would make your code pay another time the price of the Mutex (because in my opinion it's still needed on subscribe to avoid API misusage)
- It could mislead users to believe it's something necessary

I still don't understand what makes your library unable to interop with this without `enable_isr_service()` being public. As long as you get a `GpioPin<Subscribed>` you should have the guarantee that initialization was done correctly.

> In addition to losing those two things, there's now two `Box`es that I have to pay for and a `Mutex` I have to hit for every new subscription.

- You still need two `Boxes` if you want to use a Closure which in my opinion will be quite a common use case.
- I can swap the Mutex to an AtomicBool but with the price of needing to handle an edge case discussed above.

Reply:
I'm not planning to use this subscribe API because I think the pin state should be decoupled from the handler. Or at least not coupled in this way. Something about having all this global state that I have little control over its lifetime makes me uncomfortable.

However if I go my own way, I still need to know that the service has been created when someone is using this and my driver.
I will be avoiding the mutex per subscription by using a type to represent the call to enable the service. So as long as the user has this type, I know statically that it's been called and can create a handler without worry. Just like what I have in the gist.

Users thinking they might need to call it when they don't is not great, but if they are using a driver written in C++, they might need to call it first before using the driver. Otherwise when they use this, it too will try and create it's own service and fail. This might be very uncommon though.

Reply:
Thank you for your feedback.

If you go your own way, you can still use a call to the unsafe esp_idf_sys underlying function or even provide a crate with all the required wrapped implementations.

I was proposing this as a PR to this project because adding a TypeState can only be done from within this crate.

@ivmarkov @MabezDev if you want we can hide this behind a specific feature. I see value in this for me and thought it could be usefull for others but I don't want to impose anything here.